docs/Configuration.md
# CONFIGURATION

Configuration des agendas du praticien : 

![Menu configuration](images/menu-configuration.png)

Dans cette page, plusieurs options sont disponibles.

Tout d’abord, il est possible de définir un nom pour le planning. 
Pour cela, il suffit de le saisir dans le champ « Titre » :

![alt text](https://image.noelshack.com/fichiers/2018/15/5/1523621170-capture-d-ecran-2018-04-10-a-16-35-16.png)


On pourra également ajouter une description au Planning ainsi qu'une couleur par défaut pour les rendez-vous.

Toute modification effectuée dans le menu de configuration devra être enregistrée via le bouton "**Sauvegarder**" en bas de page.

## Utilisateurs et droits d'accès

Permet de définir les propriétaires et l'administrateur du planning sélectionné.

L'administrateur est unique et il ne peut donc y en avoir qu'un.
C'est lui qui définit quels sont les propriétaires d'un agenda. C'est à dire ceux ayant la possibilité d'accéder au planning et de pouvoir gérer celui-ci.
L'administrateur peut ajouter autant de propriétaires qu'il le souhaite en passant par le menu de configuration.
Pour ajouter un nouveau propriétaire, il suffit de saisir son nom dans la liste. 
Au cas où il y aurait plusieurs propriétaires, il faudra séparer chaque propriétaire par un espace.

**Exemple** : 

![Plusieurs propriétaires](https://image.noelshack.com/fichiers/2018/15/5/1523621730-capture-d-ecran-2018-04-13-a-14-14-36.png)


Il est cependant possible de changer d'administrateur à tout moment.
Pour transférer les droits d'administration à un autre utilisateur, il suffit de saisir son nom dans le champ "Transfert des droits administrateur" et de cliquer ensuite sur "Transférer" ou "Sauvegarder" :

![Nouvel administrateur](https://image.noelshack.com/fichiers/2018/15/5/1523621959-capture-d-ecran-2018-04-13-a-14-18-45.png)


A partir de là, l'utilisateur indiqué dans "Transfert des droits administrateur" verra sur sa page un bouton "Transférer un planning :

![Transférer un planning](https://image.noelshack.com/fichiers/2018/15/5/1523622193-capture-d-ecran-2018-04-13-a-14-21-49.png)



Il n'aura donc plus qu'à cliquer sur ce bouton (qui lui demandera ensuite de confirmer ou non la prise de droits) pour devenir le nouvel administrateur :

![Transfer de planning confirmation](https://image.noelshack.com/fichiers/2018/15/5/1523622193-capture-d-ecran-2018-04-13-a-14-22-02.png)



Enfin, le message suivant apparaitra pour confirmer la réussite du transfert de droits :

![Message de confirmation transfert de planning](https://image.noelshack.com/fichiers/2018/15/5/1523622193-capture-d-ecran-2018-04-13-a-14-22-10.png)



Tant que la personne à qui les droits sont transférés n'a pas confirmer la récupération des droits administrateurs, c'est celui qui a initié le transfert qui restera administrateur. 
Cela dans le but d'éviter de perdre les droits administrateur en cas d'erreur dans le nom saisi pour le transfert. 



## Plages horaires d'ouverture

Ce menu permet la configuration complète des plages d'ouverture du cabinet.

Trois paramètres sont modifiables, indépendament des journées :
La "Durée par défaut d'un RDV", qui correspond au temps par défaut pour le créneau d'un nouveau rendez-vous, et la "Durée minimale d'un RDV", qui correspond à la durée visible des créneaux horaires dans l'agenda.
Le troisième paramètre est la couleur, qui sera ainsi la couleur par défaut des rendez-vous.

Une fois ces paramètres saisis, il est possible de configurer les plages horaires d'ouverture soit pour la semaine entière directement, soit journée par journée.

Pour la partie semaine :
à compléter.

Pour configurer journée par journée, il suffit de cliquer sur le jour souhaité puis de cliquer sur le bouton "**+**", afin de définir les plages horaires souhaitées.
Il est par exemple possible d'ouvrir le lundi, de 9h à 12h et de 14h à 18h :

![Plages horaires](images/Plages-horaires.png)


Pour supprimer une plage horaire, il faut simplement cliquer sur le bouton "**-**" en bout de ligne.

Une fois toutes les plages configurées, il ne reste plus qu'à tout valider via le bouton "**Sauvegarder**" en bas de page.


## Evènements récurrents

Il est possible de créer des évenènements récurrents qui apparaitront ensuite dans le planning et qui permettront d'indiquer ce qu'on a prévu de faire à tel jour et tel créneau horaire.
Pour créer un évènement récurrent, il suffit de cliquer sur le bouton "**+ Ajouter un évènement**" :

![Evenement recurrents](images/evenement-recurrent.png)

Il faut ensuite donner un nom à l'évènement, lui choisir une couleur et indiquer si cette couleur apparaitra en arrière plan ou non.
Arrière plan signifie que la couleur sera beaucoup plus transparente. Sans l'option arrière plan, la couleur sera de la même intensité que celles des rendez-vous.

**Exemple sans l'option arrière plan** :

![Evenement recurrent arriere plan on](images/evenement-arriere-plan-on.png)


**Exemple avec l'option arrière plan** :

![Evenement recurrent arriere plan off](images/evenement-arriere-plan-off.png)

Une fois ces paramètres validés, il faut maintenant indiquer les créneaux horaires concernés par l'évènement.
Le fonctionnement est le même que pour les plages horaires d'ouverture : on sélectionne une journée, on clique sur le bouton "**+**" et on mentionne le créneau horaire souhaité.

**Exemple** : 

![alt text](images/exemple-evenement-recurrent.png)

Toutefois, il est possible d'indiquer qu'un évènement s'applique à une journée entière plutôt qu'un créneau horaire.
Pour cela, il faut cocher l'option "Toute la journée" :

![Configuration evenement recurrent](images/evenement-journee-config.png)

L'évènement récurrent s'affichera ainsi dans la zone "Toute la journée" de l'agenda :

![Evenement recurrent sur l'Agenda](images/evenement-journee-agenda.png)


## Prise de rendez-vous

La gestion des congés, des niveaux d'autorisation pour la prise des rendez-vous ainsi que la gestion des motifs de rendez-vous se fait via ce menu.
Ici, on peut définir le délai maximum, en jours, pour la prise d'un rendez-vous ainsi que le délai de prévenance, en heures, pour leur annulation.
On peut également définir le format des noms affichés dans l'agenda, comme **NOM + PRENOM**, **NOM + Prénom**, **Nom + Prénom**, etc. et indiquer la dénomination d'un rendez-vous par défaut pour les patients non identifiés :

![Configuration prise de rendez-vous](images/prise-rdv.png)


### Congés

Tout d'abord, il est possible d'activer, ou non, l'affichage des jours fériés légaux dans l'agenda.
Si on souhaite les afficher, il suffit de cocher la case "Les jours fériés légaux sont pris en compte automatiquement" :

![Les jours fériés légaux sont pris en compte](images/jours-feries.png)


Pour obtenir le résultat suivant :

![Jour férié sur l'Agenda](images/exemple-jour-ferie.png)


Ensuite, on indique si on souhaite afficher les congés dans l'agenda, la couleur de ceux-ci et enfin les dates concernées.
Là encore, on clique sur le bouton "**+**" pour ajouter une plage de dates, on saisit ensuite les dates et enfin son intitulé :

![Configuration des congés](images/conges-config.png)


### Niveaux d'autorisation

Les niveaux d'autorisation, qui vont de 0 à 3, permettent de définir ce qu'un patient ou un correspondant peut faire au niveau du planning.
Par exemple, si le patient **A** peut prendre des rendez-vous quand il le souhaite, ou bien uniquement pour certains motifs, ou encore certains créneaux horaires uniquement.
Avant de pouvoir indiquer ce qu'un patient ou un correspondant peut faire, il faut donc commencer par définir ce qu'il est possible de faire selon chaque niveau d'autorisation.

Le "**Niveau minimum d'autorisation requis pour la prise de RDV en ligne**" permet de savoir à partir de quel niveau on peut accéder aux rendez-vous en ligne :

![Niveau minimum de prise de rendez-vous](images/niveau-minimum.png)

Il est possible de désactiver cette fonctionnalité en sélectionnant la ligne "RDV en ligne désactivés (4)".

Le "**Niveau maximum accepté pour un RDV en ligne**" permet cette fois d'indiquer le niveau maximum autorisé (A REVOIR).

![Niveau maximum de prise de rendez-vous](images/niveau-maximum.png)


Le "**Niveau d'autorisation minimum des motifs proposés pour la prise de RDV depuis l'agenda**" sert à définir à partir de quel niveau on peut sélectionner un motif lors de notre prise de rendez-vous en ligne.

![Niveau motifs](images/niveau-motifs.png)


Enfin, les "**Plages horaires ouvertes par niveau d'autorisation**" sont à définir de la même manière que les plages horaires classiques.
Il faut d'abord sélectionner le niveau d'autorisation pour lequel on veut définir les plages, choisir ensuite une journée et cliquer sur le bouton "**+**" pour sélectionner le créneau souhaité.

Par exemple, si on veut indiquer que les personnes, dont le niveau d'autorisation est 0, ont le droit de prendre des rendez-vous en ligne le lundi de 8h à 12h, on sélectionne la rubrique "**Niveau d'autorisation 0**", on sélectionne le lundi et on ajoute la plage 8h-12h :

![Plage niveau 0](images/plage-niveau0.png)



### Motifs des rendez-vous

Il est possible de préparer des motifs de rendez-vous afin de ne pas avoir à les saisir manuellement pour chaque rendez-vous futur.

Il faut utiliser le bouton "**+**" pour créer un nouveau motif et ensuite, saisir son intitulé, son niveau d'autorisation (pour la prise de rendez-vous en ligne), sa durée par défaut (qui pourra bien entendu être modifiée lors de la prise d'un rendez-vous) et sa couleur :

![Motifs des rendez-vous](images/motif-rdv.png)




## Rappels

Ce menu permet au praticien de définir s'il souhaite ou non envoyer des rappels de SMS aux patients.
Pour activer l'option de rappels par SMS, il suffit de cocher la case "Emettre des rappels par SMS"

![Illustration rappel SMS 1](images/rappels-rdv-sms-2.png)


Une fois l'option activée, il faudra bien veiller à cocher les délais de rappels qui sont tous activés par défaut.
Ainsi, pour activer uniquement un rappel toutes les 48h, il ne faudra cocher que la case "48h avant le rendez-vous".

**Exemple** :

![Illustration rappel SMS 2](images/rappels-rdv-48heures.png)


Il est ensuite possible de personnaliser les deux types de SMS envoyés aux patients :

1) Le SMS de confirmation initiale. C'est le SMS envoyé au patient lorsqu'il prend rendez-vous en ligne pour qu'il puisse le confirmer :

![Texte de confirmation SMS](images/texte-confirmation-sms.png)


2) Le SMS de rappel, celui envoyé 48h, 24h ou 1h (selon le(s) choix effectué(s)) au patient pour lui rappeler son rendez-vous :

![Texte de rappel par SMS](images/texte-rappel-sms.png)

Dans ces deux zones de texte, tout ce qui est écrit entre les symboles **{}** sont des champs automatiques.
Par exemple, le champ **{date-heure}** retourne la date et l'heure du rendez-vous du patient, dans le SMS.
Il ne faut donc pas modifier ces types de champs.
Tout le reste par contre est modifiable.


## Réutilisation des configurations

![Réutilisation des configurations](images/reutilisation-des-configurations.png)

* Le praticien peut **Sauvegarder** la configuration en cours et elle sera enregistré dans un fichier au format **JSON** nommé "**config-plannings.json**" et placé dans le dossier des **Téléchargements**.

* Le praticien peut réstaurer une configuration. Il faut **charger une configuration** à partir d'une sauvegarde et tous les plannings seront remplacés par les plannings de ce fichier. Si ce fichier est conforme à la configuration courante, cette restauration préservera les différents états des rendez-vous déjà pris.

## Import export au format iCalendar (.ics)

![Import export des rendez-vous](images/import-export-rdv.png)

* Le praticien peut **Exporter** tous les rendez-vous se trouvant sur un planning sélectionné. Ces derniers seront dans un fichier au format **iCalendar** nommé "**_nom-planning.ics_**" et placé dans le dossier des **Téléchargements**.

* Il est possible d'importer des rendez-vous à partir d'un fichier au format **iCalendar**. Par défaut, ces rendez-vous seront importés sur le planning sélectionné et ceux préexistants sur celui-ci seront supprimés. Pour ne pas supprimer les rendez-vous préexistants sur un planning, il faut désactiver l'option **_Effacer tous les rendez-vous préexistants sur le(s) planning(s) de destination_**. Pour choisir d'autres plannings sur lesquels importer des rendez-vous, il faut activer l'option **_Importer les rendez-vous sur d'autres plannings_**.


src/components/Settings.js
import site from "./SiteSettings";

import moment from "moment";

import _ from "lodash";

/*
 * maxWidth : Largeur de la colonne centrale
 * (d'autres largeurs sont définies
 *  comme des fractions de maxWidth)
 */

const maxWidth = 500;

/*
 * psize : Preferred Semantic UI Form size
 * hsize : Preferred Semantic UI Header/Title Size
 */

const fsize = "large";
const hsize = "large";

/*
 * Affichage par défaut Date/Heure d'un rendez-vous (format long en Français)
 */

const rdvDateTime = dateStr =>
  moment(dateStr).format("dddd D MMMM YYYY à HH:mm");

/*
 * defaultPlanning : Planning par défaut
 */

const defaultPlanning = {
  titre: "Sans titre",
  description: "Nouveau planning",
  couleur: "#7ED321",
  optionsJO: {
    acl: {
      owners: [],
      admin: ""
    },
    plages: {
      duree: 30,
      dureeMin: 15,
      horaires: [
        [],
        [
          {
            start: "09:00",
            end: "12:00"
          },
          {
            start: "14:00",
            end: "19:00"
          }
        ],
        [
          {
            start: "09:00",
            end: "12:00"
          },
          {
            start: "14:00",
            end: "19:00"
          }
        ],
        [
          {
            start: "09:00",
            end: "12:00"
          },
          {
            start: "14:00",
            end: "19:00"
          }
        ],
        [],
        [
          {
            start: "09:00",
            end: "12:00"
          },
          {
            start: "14:00",
            end: "19:00"
          }
        ],
        [
          {
            start: "09:00",
            end: "12:00"
          }
        ]
      ]
    },
    recurrents: [
      {
        titre: "Réunion début de semaine",
        couleur: "#000000",
        background: true,
        recurrence: 0,
        from: 0,
        step: 0,
        start: "",
        end: "",
        horaires: [
          [],
          [
            {
              start: "09:00",
              end: "10:00"
            }
          ],
          [],
          [],
          [],
          [],
          []
        ]
      }
    ],
    reservation: {
      autorisationMin: 0,
      autorisationMax: 2,
      autorisationMinAgenda: 1,
      congesCouleur: "#D0021B",
      congesVisibles: true,
      congesFeries: true,
      conges: [
        {
          titre: "Fermeture annuelle",
          start: "2018-08-01",
          end: "2018-08-20"
        },
        {
          titre: "Congés de fin d'année",
          start: "2018-12-20",
          end: "2019-01-02"
        }
      ],
      delaiMax: 120,
      delaiPrevenance: 48,
      denominationFormat: "NP",
      denominationDefaut: "Anonyme",
      planningsAssocies: [],
      motifs: [
        {
          autorisationMin: 0,
          couleur: "#4A90E2",
          duree: 10,
          motif: "Première consultation"
        },
        {
          autorisationMin: 0,
          couleur: "#4A90E2",
          duree: 10,
          motif: "Je suis adressé(e) pour avis par un médecin/un dentiste"
        },
        {
          autorisationMin: 1,
          couleur: "#4A90E2",
          duree: 15,
          motif: "Consultation de suivi"
        },
        {
          autorisationMin: 2,
          couleur: "#4A90E2",
          duree: 30,
          motif: "Intervention programmée"
        }
      ],
      horaires: [
        [
          [],
          [],
          [
            {
              start: "14:00",
              end: "19:00"
            }
          ],
          [],
          [],
          [],
          []
        ],
        [
          [],
          [
            {
              start: "09:00",
              end: "12:00"
            },
            {
              start: "14:00",
              end: "19:00"
            }
          ],
          [
            {
              start: "09:00",
              end: "12:00"
            },
            {
              start: "14:00",
              end: "19:00"
            }
          ],
          [
            {
              start: "09:00",
              end: "12:00"
            },
            {
              start: "14:00",
              end: "19:00"
            }
          ],
          [],
          [
            {
              start: "09:00",
              end: "12:00"
            },
            {
              start: "14:00",
              end: "19:00"
            }
          ],
          []
        ],
        [[], [], [], [], [], [], []],
        [[], [], [], [], [], [], []]
      ]
    }
  },
  sms: {
    rappel1: false,
    rappel24: false,
    rappel48: false,
    rappelTexte:
      "Bonjour,\nRDV {date-heure}.\nDr (saisir les coordonnées)\n{infos-annulation}",
    confirmationTexte:
      "RDV {date-heure}.\nDr (saisir les coordonnées)\n{infos-annulation}"
  }
};

/*
 * Regex code postal, email, et téléphones
 */
const codePostalRegex = /^[0-9]{5}$/;

const emailRegex = /(^\w)([\w+.-])*([\w+-])*(@)([\w+.-])+\.([a-z]{2,4})$/i;

//Les formats valides :
//  01 12 45 78 14 pattern 1 (ou sans espace)
//  01.12.45.78.14 pattern 2
//  +33 6 00 00 00 00 pattern 3 (ou sans espace)
//  00 33 6 00 00 00 00 pattern 4 (ou sans espace)
const telRegex = [
  /^0[1-9]([\s.]?[0-9]{2}){4}$/, // pattern 1 et 2
  /^\+[1-9][0-9]{1,2}(\s)?[1-9](\s?[0-9]{2}){4}$/, // pattern 3
  /^00(\s)?[1-9][0-9]{1,2}(\s)?([1-9])(\s?[0-9]{2}){4}$/
];

// Formatage du numéro de téléphone pour l'affichage seulement
const telFormat = telephone => {
  if (telephone === "") {
    return "";
  }
  let result = "";
  for (let i = 0; i < telRegex.length; i++) {
    if (i === 0 && telRegex[i].test(telephone)) {
      // match avec les patterns 1 et 2
      let val = telephone.replace(/(\.|\s)/g, ""); // remplacer tous les . par des espaces
      for (let j = 0; j < val.length; j++) {
        if (j % 2 === 0 || j === val.length - 1) {
          result += val[j];
        } else {
          result = result + val[j] + " ";
        }
      }
      return result;
    } else if (i === 1 && telRegex[i].test(telephone)) {
      // match avec le pattern 3
      let val = telephone.replace(/\s/g, "");
      if (val.length === 12) {
        for (let j = 0; j < val.length; j++) {
          if ((j === 1 || j === val.length - 1 || j % 2 === 0) && j !== 2) {
            result += val[j];
          } else {
            result = result + val[j] + " ";
          }
        }
        return result;
      } else if (val.length === 13) {
        for (let j = 0; j < val.length; j++) {
          if (
            j === 0 ||
            j === 2 ||
            j === val.length - 1 ||
            (j % 2 !== 0 && j !== 3)
          ) {
            result += val[j];
          } else {
            result = result + val[j] + " ";
          }
        }
        return result;
      }
    } else if (i === 2 && telRegex[i].test(telephone)) {
      // match avec le pattern 4
      let val = telephone.replace(/\s/g, "");
      if (val.length === 13) {
        for (let j = 0; j < val.length; j++) {
          if (
            j === 0 ||
            j === 2 ||
            j === val.length - 1 ||
            (j % 2 !== 0 && j !== 1 && j !== 3)
          ) {
            result += val[j];
          } else {
            result = result + val[j] + " ";
          }
        }
        return result;
      } else if (val.length === 14) {
        for (let j = 0; j < val.length; j++) {
          if (j === 4 || (j % 2 !== 0 && j !== 3 && j !== val.length - 1)) {
            result = result + val[j] + " ";
          } else {
            result += val[j];
          }
        }
        return result;
      }
    }
  }
};

/*
 * Format de dénomination
 */
const denominationDefaultFormat = "NP";

/*
 * Fonctions qui seront utilisées pour le formatage de la dénomination
 * lors de l'affichage
 * -> camelDenomination et affichageDenomination
 */
const camelDenomination = text => {
  let result = "";
  let prev = "";
  for (let i = 0; i < text.length; i++) {
    let c = text[i];
    if (i === 0 || prev === " " || prev === "'" || prev === "-") {
      c = _.toUpper(c);
    } else {
      c = _.toLower(c);
    }
    prev = c;
    result += c;
  }
  return result;
};

const affichageDenomination = (denominationDefaultFormat, nom, prenom) => {
  switch (denominationDefaultFormat) {
    case "NP":
      return _.toUpper(nom) + " " + _.toUpper(prenom);
    case "Np":
      return _.toUpper(nom) + " " + camelDenomination(prenom);
    case "PN":
      return _.toUpper(prenom) + " " + _.toUpper(nom);
    case "pN":
      return camelDenomination(prenom) + " " + _.toUpper(nom);
    case "np":
      return camelDenomination(nom) + " " + camelDenomination(prenom);
    case "pn":
      return camelDenomination(prenom) + " " + camelDenomination(nom);
    default:
      return nom + " " + prenom;
  }
};

// Retourne une chaine de caractères décrivant la civilité
// ex : M., Mme, Professeur etc.
const civilite = (short, valCivilite) => {
  let civilites = [
    { text: "", shorttext: "", value: 0 },
    { text: "Monsieur", shorttext: "M.", value: 1 },
    { text: "Madame", shorttext: "Mme", value: 2, hidden: true },
    { text: "Mademoiselle", shorttext: "Mlle", value: 3 },
    { text: "Enfant", shorttext: "Enfant", value: 4 }
  ];

  let civiliteNum = 1 * valCivilite;
  let civiliteStr = "" + valCivilite;
  if (!isNaN(civiliteNum)) {
    if (civiliteNum < civilites.length) {
      civiliteStr = short
        ? civilites[civiliteNum].shorttext
        : civiliteNum === 3 // Mademoiselle (obsolète) est géré comme un texte libre (autre)
        ? civilites[civiliteNum].text
        : "";
    } else {
      civiliteStr = "";
    }
  }
  return civiliteStr;
};

/*
 * rdvEtats états des rendez-vous (idEtat du groupe RendezVous)
 */
const rdvEtats = [
  { text: "RDV non confirmé", color: "black" }, // (0) RDV pris en ligne non encore confirmé (masqué)
  { text: "Statut non renseigné", color: "lightgrey" }, // (1)
  { text: "Présence à l'heure", color: "lime" }, // (2)
  { text: "Retard", color: "yellow" }, // (3)
  { text: "Retard important", color: "orange" }, // (4)
  { text: "RDV annulé dernier moment", color: "red" }, // (5)// Hors délai de prévenance
  { text: "Absence", color: "red" }, // (6)
  { text: "RDV annulé", color: "black" } // (7) // Annulation dans les délais (masqué)
];

/*
 * popups d'aide
 */
const helpPopup = {
  on: "hover",
  size: "mini",
  inverted: true
};

// local dev => no auth -> identification patient par id + password
const localdev = false;

export {
  localdev,
  site,
  maxWidth,
  helpPopup,
  fsize,
  hsize,
  defaultPlanning,
  rdvDateTime,
  codePostalRegex,
  emailRegex,
  telRegex,
  telFormat,
  civilite,
  camelDenomination,
  denominationDefaultFormat,
  affichageDenomination,
  rdvEtats
};
